fix some codegen bugs: NIM_BOOL, NIM_STATIC_ASSERT, --passc:-std=... (etc)

timotheecour opened the pull request:
* `--passc:-std=c++17` (or similar flags) are now honored instead of silently ignored, by putting `--passc` options **after** instead of **before** the default
* use correct definition for `NIM_BOOL ` in C target. Previous definition was incorrect, as `bool` is not guaranteed to be compatible with `unsigned char` (could have different sizeof etc); furthermore, it would cause codegen error for C code that uses `_Bool` or (after `include <stdbool.h>`) `bool`
note: it's >= C99 but I doubt nim codegen works with < C99; if needed an extra check (or conditional define) can be used

* `NIM_NIL` is incorrect for c++ < c++11, but fixing it would be hard and maybe not worth it, eg:
```nim
when true:
  {.emit:"""
  #include <stdio.h>
  struct Foo {
    int fun(void* a) { return 1;}
    int fun(int a){ return 2; }
  };
  """.}
  type Foo {.importcpp.} = object
  proc fun(foo: var Foo, a: pointer): cint {.importcpp.}
  var foo: Foo
  echo foo.fun(nil)
```

```
nim cpp -r --passc=-std=c++03 $timn_D/tests/nim/all/t10447.nim
2
nim cpp -r --passc=-std=c++17 $timn_D/tests/nim/all/t10447.nim
1
```
(requires rebuilding nim with this PR so that `--passc=-std=c++03` is honored)

## EDIT
* fixes https://github.com/nim-lang/Nim/issues/13933

## links
* https://stackoverflow.com/questions/1608318/is-bool-a-native-c-type/1608350
* D20200328T203812

timotheecour committed:
fix cgen bool D20200328T203812

timotheecour committed:
--passc:std=c++17 (etc) now works instead of silently ignored

timotheecour committed:
document caveats for NIM_NIL

timotheecour committed:
add test

timotheecour committed:
fixup

timotheecour committed:
fix test

cooldome commented on compiler/extccomp.nim:
```diff
@@ -612,6 +612,11 @@ proc getCompileCFileCmd*(conf: ConfigRef; cfile: Cfile,
   var options = cFileSpecificOptions(conf, cfile.nimname, cfile.cname.changeFileExt("").string)
+  if useCpp(conf, cfile.cname):
+    # needs to be prepended so that --passc:-std=c++17 can override
+    # we could avoid allocatoin by making cFileSpecificOptions inplace
```
typo: allocatoin

timotheecour replied:
fixed

Araq commented:
``stdbool.h`` is useless though as it doesn't say anything about ``sizeof(bool)`` and Nim requires it to be 1 byte.

timotheecour committed:
NIM_STATIC_ASSERT, trunner

timotheecour committed:
fixup

timotheecour commented:
PTAL

> stdbool.h is useless though as it doesn't say anything about sizeof(bool) and Nim requires it to be 1 byte.

* removed `stdbool.h` and used `_Bool` directly instead to defined `NIM_BOOL`, which still fixes issue in top post
* I'm defining a portable `NIM_STATIC_ASSERT` and checking `NIM_STATIC_ASSERT(sizeof(bool)==1)`; if a weird platform shows up where `sizeof(_Bool)!=1`, we'll have a CT error instead of potentially weird bugs and we'll adjust nimbase.h accordingly (eg via `#ifdef NIM_DEFINE_BOOL_AS_UCHAR`)
*  `NIM_STATIC_ASSERT` is better (and easier to reuse) than the old hack as it shows line number and failing condition; I added a test for checking its behavior

* `NIM_STATIC_ASSERT` is a key ingredient to support CT sizeof for importc types, see https://github.com/nim-lang/RFCs/issues/205 for details

* couldn't figure out how to use the testament `discard` spec for a test like `tests/ccgbugs/mstatic_assert.nim` so I'm instead compiling it in `tests/trunner.nim` (after refactoring that file from `tests/vm/tevalffi.nim`); the few other tests that also need an explicit compilation step (or non-trivial checks on the output) can also be simplified by merging their logic into `trunner` (better than making testament spec more complex with more magic flags IMO)

* related to that, I've fixed a bug so that CTFFI now tests the c++ (instead of C) target depending on `NIM_COMPILE_TO_CPP`

Araq commented:
> it's >= C99 but I doubt nim codegen works with < C99; if needed an extra check (or conditional define) can be used

We support C89.

timotheecour committed:
support stdc < C99

timotheecour commented:
> We support C89.

done, that was a simple fix, PTAL

Araq commented on lib/nimbase.h:
```diff
@@ -288,34 +303,40 @@ __AVR__
+// see #13798: to avoid conflicts for code emitting `#include <stdbool.h>`
+#define NIM_BOOL _Bool
+#else
+typedef unsigned char NIM_BOOL // best effort
```
Missing ';'

timotheecour replied:
nice catch, fixed

timotheecour committed:
fixup

timotheecour commented:
friendly ping @Araq
